File: NotePal/Sources/NotePal/Models/TodoItem.swift
import Foundation

struct TodoItem: Identifiable, Codable, Equatable {
    var id: UUID
    var title: String
    var description: String?
    var isCompleted: Bool
    var dueDate: Date?
    var wellnessKind: String?
    var wellnessReminderIntervalMinutes: Int?
    var createdAt: Date
    var updatedAt: Date

    init(
        id: UUID = UUID(),
        title: String = "新待办",
        description: String? = nil,
        isCompleted: Bool = false,
        dueDate: Date? = nil,
        wellnessKind: String? = nil,
        wellnessReminderIntervalMinutes: Int? = nil,
        createdAt: Date = Date(),
        updatedAt: Date = Date()
    ) {
        self.id = id
        self.title = title
        self.description = description
        self.isCompleted = isCompleted
        self.dueDate = dueDate
        self.wellnessKind = wellnessKind
        self.wellnessReminderIntervalMinutes = wellnessReminderIntervalMinutes
        self.createdAt = createdAt
        self.updatedAt = updatedAt
    }

    var isWellnessTodo: Bool {
        wellnessKind != nil
    }

    var effectiveWellnessIntervalMinutes: Int {
        max(5, wellnessReminderIntervalMinutes ?? 60)
    }

    var reminderRevisionKey: String {
        let dueStamp = dueDate.map { String(Int($0.timeIntervalSince1970)) } ?? "none"
        let updatedStamp = String(Int(updatedAt.timeIntervalSince1970 * 1000))
        guard isWellnessTodo else {
            return "v2|\(id.uuidString)|\(dueStamp)|\(updatedStamp)"
        }

        let wellnessStamp = wellnessKind ?? "regular"
        let intervalStamp = wellnessReminderIntervalMinutes.map(String.init) ?? "none"
        return "v3|\(id.uuidString)|\(dueStamp)|\(updatedStamp)|\(wellnessStamp)|\(intervalStamp)"
    }

    func isDue(asOf date: Date) -> Bool {
        guard let dueDate, !isCompleted else {
            return false
        }

        return dueDate <= date
    }
}

File: NotePal/Windows/NotePal.Windows/Models/TodoItem.cs
using System.Globalization;

namespace NotePalWindows.Models;

public sealed class TodoItem
{
    public Guid Id { get; set; } = Guid.NewGuid();
    public string Title { get; set; } = "新待办";
    public string? Description { get; set; }
    public bool IsCompleted { get; set; }
    public DateTimeOffset? DueDate { get; set; }
    public string? WellnessKind { get; set; }
    public int? WellnessReminderIntervalMinutes { get; set; }
    public DateTimeOffset CreatedAt { get; set; } = DateTimeOffset.Now;
    public DateTimeOffset UpdatedAt { get; set; } = DateTimeOffset.Now;

    public bool IsWellnessTodo => WellnessKind is not null;

    public int EffectiveWellnessIntervalMinutes => Math.Max(5, WellnessReminderIntervalMinutes ?? 60);

    public string ReminderRevisionKey
    {
        get
        {
            var dueStamp = DueDate?.ToUnixTimeSeconds().ToString(CultureInfo.InvariantCulture) ?? "none";
            var updatedStamp = UpdatedAt.ToUnixTimeMilliseconds().ToString(CultureInfo.InvariantCulture);
            if (!IsWellnessTodo)
            {
                return $"v2|{Id.ToString().ToUpperInvariant()}|{dueStamp}|{updatedStamp}";
            }

            var wellnessStamp = WellnessKind ?? "regular";
            var intervalStamp = WellnessReminderIntervalMinutes?.ToString(CultureInfo.InvariantCulture) ?? "none";
            return $"v3|{Id.ToString().ToUpperInvariant()}|{dueStamp}|{updatedStamp}|{wellnessStamp}|{intervalStamp}";
        }
    }

    public bool IsDue(DateTimeOffset asOf)
    {
        return DueDate is not null && !IsCompleted && DueDate <= asOf;
    }

    public TodoItem Clone()
    {
        return new TodoItem
        {
            Id = Id,
            Title = Title,
            Description = Description,
            IsCompleted = IsCompleted,
            DueDate = DueDate,
            WellnessKind = WellnessKind,
            WellnessReminderIntervalMinutes = WellnessReminderIntervalMinutes,
            CreatedAt = CreatedAt,
            UpdatedAt = UpdatedAt
        };
    }
}
